base/utf.jl

```diff
@@ -0,0 +1,969 @@
+const UTF_ERR_INVALID = 8
+const UTF_ERR_SURROGATE = 9
+const UTF_ERR_NULL_16_TERMINATE = 10
+const UTF_ERR_NULL_32_TERMINATE = 11
+const UTF_ERR_MAX = 11
+
+const errMsgs = [
+    "invalid UTF-8 sequence starting at index <<1>> (0x<<2>>) missing one or more continuation bytes)",
+    "invalid UTF-8 sequence starting at index <<1>> (0x<<2>> is not a continuation byte)",
+    "invalid UTF-8 sequence, overlong encoding starting at index <<1>> (0x<<2>>)",
+    "not a leading Unicode surrogate character at index <<1>> (0x<<2>>)",
+    "not a trailing Unicode surrogate character at index <<1>> (0x<<2>>)",
+    "not a valid Unicode surrogate character at index <<1>> (0x<<2>>",
+    "missing trailing Unicode surrogate character after index <<1>> (0x<<2>>)",
+    "invalid Unicode character starting at index <<1>> (0x<<2>> > 0x10ffff)",
+    "surrogate encoding not allowed in UTF-8 or UTF-32, at index <<1>> (0x<<2>>)",
+    "UTF16String data must be NULL-terminated",
+    "UTF32String data must be NULL-terminated"
+]
+#=
+@doc """
+@brief      Throws ArgumentError with information about the specific error, location, and character
+
+@param[in]  errcode Error code for Unicode error (one of UTF_ERR_*)
+@param[in]  charpos Index of invalid byte or character
+@param[in]  invchar Invalid byte or character
+
+@throws never returns, always throws ArgumentError
+""" ->
+=#
+function utf_errfunc(errcode::Integer, charpos, invchar)
+    if errcode < 1 || errcode > UTF_ERR_MAX
+        throw(ArgumentError("Invalid error code for Unicode error: $errcode, Pos = $charpos, Char = $invchar"))
+    end
+    throw(ArgumentError(replace(replace(errMsgs[errcode],"<<1>>",string(charpos)),"<<2>>",hex(invchar))))
+end
+
+#=
+@doc """
+@brief      Base UTF16String type, has 16-bit NULL termination word after data, native byte order
+""" ->
+=#
+immutable UTF16String <: AbstractString
+    data::Vector{UInt16} # includes 16-bit NULL termination after string chars
+    function UTF16String(data::Vector{UInt16})
+        if length(data) < 1 || data[end] != 0
+            utf_errfunc(UTF_ERR_NULL_16_TERMINATE, 0, 0)
+        end
+        new(data)
+    end
+end
+
+#=
+@doc """
+@brief      Base UTF32String type, has 32-bit NULL termination word after data, native byte order
+""" ->
+=#
+immutable UTF32String <: DirectIndexString
+    data::Vector{Char} # includes 32-bit NULL termination after string chars
+
+    function UTF32String(data::Vector{Char})
+        if length(data) < 1 || data[end] != Char(0)
+            utf_errfunc(UTF_ERR_NULL_32_TERMINATE, 0, 0)
+        end
+        new(data)
+    end
+end
+UTF32String(data::Vector{UInt32}) = UTF32String(reinterpret(Char, data))
+
+const empty_utf16 = UTF16String(UInt16[0])
+const empty_utf32 = UTF32String(UInt32[0])
+
+is_surrogate_lead(c::Unsigned) = ((c & ~0x003ff) == 0xd800)
+is_surrogate_trail(c::Unsigned) = ((c & ~0x003ff) == 0xdc00)
+is_surrogate_char(c::Unsigned) = ((c & ~0x007ff) == 0xd800)
```

Comment:
I removed the `utf16_`, because I am using these also for surrogates encoded in `UTF-8` and `UTF-32` (i.e. for `CESU-8` strings)
About is_surrogate_char, I can see your point... 2 things - I dislike having names that are substrings of other names, *esp* in the same file, makes things hard to find or do regex changes...
I was thinking about the `_char` in terms of Char, which currently is just a codepoint, not even validatated, but maybe `is_surrogate_codepoint` would be better?